File: 6.actions/src/action_py_pkg/action_py_pkg/turtle_action_client.py
#!/usr/bin/env python3
import rclpy
from action_msgs.msg import GoalStatus
from rclpy.action import ActionClient
from rclpy.node import Node

from custom_interfaces.action import MoveToGoal


class TurtleActionClient(Node):
    """Action client that sends navigation goals to turtle_action_server and monitors progress."""

    def __init__(self):
        super().__init__("turtle_action_client")

        # Declare parameters so users can customize goals via CLI or launch
        self.declare_parameter("target_x", 8.5)
        self.declare_parameter("target_y", 8.5)
        self.declare_parameter("linear_velocity", 2.0)
        self.declare_parameter("cancel_after_sec", 0.0)

        # Create action client
        self.client = ActionClient(self, MoveToGoal, "move_to_goal")
        self.goal_handle = None
        self.cancel_timer = None

        self.get_logger().info("Turtle Action Client has been started!")

        # Short timer to trigger sending goal after node initialization
        self.init_timer = self.create_timer(0.5, self.timer_callback)

    def timer_callback(self):
        """One-shot trigger to send the initial goal."""
        self.init_timer.cancel()
        target_x = float(self.get_parameter("target_x").value)
        target_y = float(self.get_parameter("target_y").value)
        linear_velocity = float(self.get_parameter("linear_velocity").value)
        self.send_goal(target_x, target_y, linear_velocity)

    def send_goal(self, target_x: float, target_y: float, linear_velocity: float = 2.0):
        """Send a MoveToGoal goal request to the action server."""
        self.get_logger().info("Waiting for action server '/move_to_goal'...")
        while not self.client.wait_for_server(timeout_sec=1.0):
            if not rclpy.ok():
                self.get_logger().error("Interrupted while waiting for server.")
                return
            self.get_logger().info("Action server not available, waiting again...")

        goal_msg = MoveToGoal.Goal()
        goal_msg.target_x = target_x
        goal_msg.target_y = target_y
        goal_msg.linear_velocity = linear_velocity

        self.get_logger().info(
            f"Sending goal: target=({target_x:.2f}, {target_y:.2f}), velocity={linear_velocity:.2f}"
        )

        send_goal_future = self.client.send_goal_async(
            goal_msg, feedback_callback=self.feedback_callback
        )
        send_goal_future.add_done_callback(self.goal_response_callback)

    def goal_response_callback(self, future):
        """Callback invoked when the server accepts or rejects the goal."""
        self.goal_handle = future.result()
        if not self.goal_handle.accepted:
            self.get_logger().warn("Goal was REJECTED by server!")
            return

        self.get_logger().info("Goal ACCEPTED by server! Navigating to target...")

        # If cancel_after_sec parameter is specified, schedule cancellation
        cancel_after_sec = float(self.get_parameter("cancel_after_sec").value)
        if cancel_after_sec > 0.0:
            self.get_logger().info(f"Will request cancellation in {cancel_after_sec:.1f}s...")
            self.cancel_timer = self.create_timer(cancel_after_sec, self.request_cancel)

        get_result_future = self.goal_handle.get_result_async()
        get_result_future.add_done_callback(self.result_callback)

    def feedback_callback(self, feedback_msg):
        """Callback invoked whenever the server publishes periodic feedback."""
        feedback = feedback_msg.feedback
        self.get_logger().info(
            f"Feedback: remaining distance={feedback.current_distance:.2f}m, "
            f"current_pos=({feedback.current_x:.2f}, {feedback.current_y:.2f})"
        )

    def result_callback(self, future):
        """Callback invoked when the action execution terminates."""
        result_wrapper = future.result()
        status = result_wrapper.status
        result = result_wrapper.result

        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info(
                f"Goal SUCCEEDED! Total distance: {result.total_distance:.2f}m, "
                f"Elapsed time: {result.elapsed_time:.2f}s"
            )
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().warn(
                f"Goal was CANCELED! Partial distance: {result.total_distance:.2f}m, "
                f"Elapsed time: {result.elapsed_time:.2f}s"
            )
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().warn(
                f"Goal was ABORTED (preempted by another goal)! Partial distance: {result.total_distance:.2f}m"
            )
        else:
            self.get_logger().error(f"Goal terminated with status code: {status}")

    def request_cancel(self):
        """Send a cancellation request for the active goal."""
        self.cancel_timer.cancel()
        if self.goal_handle is not None:
            self.get_logger().info("Sending cancel request to action server...")
            cancel_future = self.goal_handle.cancel_goal_async()
            cancel_future.add_done_callback(self.cancel_response_callback)

    def cancel_response_callback(self, future):
        """Callback invoked when the server responds to the cancellation request."""
        cancel_response = future.result()
        if len(cancel_response.goals_canceling) > 0:
            self.get_logger().info("Cancellation request was ACCEPTED by server.")
        else:
            self.get_logger().warn("Cancellation request was REJECTED by server.")


def main(args=None):
    rclpy.init(args=args)
    node = TurtleActionClient()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: 6.actions/src/action_cpp_pkg/src/turtle_action_client.cpp
#include "action_cpp_pkg/turtle_action_client.hpp"
#include "rclcpp_components/register_node_macro.hpp"

namespace action_cpp_pkg
{

TurtleActionClientNode::TurtleActionClientNode(const rclcpp::NodeOptions & options)
: Node("turtle_action_client", options)
{
    // Declare parameters
    this->declare_parameter<double>("target_x", 8.5);
    this->declare_parameter<double>("target_y", 8.5);
    this->declare_parameter<double>("linear_velocity", 2.0);
    this->declare_parameter<double>("cancel_after_sec", 0.0);

    // Create action client
    client_ = rclcpp_action::create_client<MoveToGoal>(this, "move_to_goal");

    RCLCPP_INFO(this->get_logger(), "Turtle Action Client component has been started!");

    // Trigger goal sending after a brief initialization delay
    init_timer_ = this->create_wall_timer(
        std::chrono::milliseconds(500),
        std::bind(&TurtleActionClientNode::timer_callback, this));
}

void TurtleActionClientNode::send_goal(float target_x, float target_y, float linear_velocity)
{
    RCLCPP_INFO(this->get_logger(), "Waiting for action server '/move_to_goal'...");
    while (!client_->wait_for_action_server(std::chrono::seconds(1)))
    {
        if (!rclcpp::ok())
        {
            RCLCPP_ERROR(this->get_logger(), "Interrupted while waiting for action server.");
            return;
        }
        RCLCPP_INFO(this->get_logger(), "Action server not available, waiting again...");
    }

    auto goal_msg = MoveToGoal::Goal();
    goal_msg.target_x = target_x;
    goal_msg.target_y = target_y;
    goal_msg.linear_velocity = linear_velocity;

    RCLCPP_INFO(this->get_logger(), "Sending goal: target=(%.2f, %.2f), velocity=%.2f",
                target_x, target_y, linear_velocity);

    auto send_goal_options = rclcpp_action::Client<MoveToGoal>::SendGoalOptions();
    send_goal_options.goal_response_callback =
        std::bind(&TurtleActionClientNode::goal_response_callback, this, std::placeholders::_1);
    send_goal_options.feedback_callback =
        std::bind(&TurtleActionClientNode::feedback_callback, this, std::placeholders::_1, std::placeholders::_2);
    send_goal_options.result_callback =
        std::bind(&TurtleActionClientNode::result_callback, this, std::placeholders::_1);

    client_->async_send_goal(goal_msg, send_goal_options);
}

void TurtleActionClientNode::timer_callback()
{
    init_timer_->cancel();
    float target_x = this->get_parameter("target_x").as_double();
    float target_y = this->get_parameter("target_y").as_double();
    float linear_velocity = this->get_parameter("linear_velocity").as_double();
    send_goal(target_x, target_y, linear_velocity);
}

void TurtleActionClientNode::goal_response_callback(const GoalHandleMoveToGoal::SharedPtr & goal_handle)
{
    if (!goal_handle)
    {
        RCLCPP_WARN(this->get_logger(), "Goal was REJECTED by server!");
        return;
    }

    RCLCPP_INFO(this->get_logger(), "Goal ACCEPTED by server! Navigating to target...");
    goal_handle_ = goal_handle;

    double cancel_after_sec = this->get_parameter("cancel_after_sec").as_double();
    if (cancel_after_sec > 0.0)
    {
        RCLCPP_INFO(this->get_logger(), "Will request cancellation in %.1fs...", cancel_after_sec);
        cancel_timer_ = this->create_wall_timer(
            std::chrono::duration<double>(cancel_after_sec),
            [this]() {
                cancel_timer_->cancel();
                if (goal_handle_)
                {
                    RCLCPP_INFO(this->get_logger(), "Sending cancel request to action server...");
                    client_->async_cancel_goal(goal_handle_);
                }
            });
    }
}

void TurtleActionClientNode::feedback_callback(
    GoalHandleMoveToGoal::SharedPtr,
    const std::shared_ptr<const MoveToGoal::Feedback> feedback)
{
    RCLCPP_INFO(
        this->get_logger(),
        "Feedback: remaining distance=%.2fm, current_pos=(%.2f, %.2f)",
        feedback->current_distance, feedback->current_x, feedback->current_y);
}

void TurtleActionClientNode::result_callback(const GoalHandleMoveToGoal::WrappedResult & result)
{
    switch (result.code)
    {
        case rclcpp_action::ResultCode::SUCCEEDED:
            RCLCPP_INFO(
                this->get_logger(),
                "Goal SUCCEEDED! Total distance: %.2fm, Elapsed time: %.2fs",
                result.result->total_distance, result.result->elapsed_time);
            break;
        case rclcpp_action::ResultCode::CANCELED:
            RCLCPP_WARN(
                this->get_logger(),
                "Goal was CANCELED! Partial distance: %.2fm, Elapsed time: %.2fs",
                result.result->total_distance, result.result->elapsed_time);
            break;
        case rclcpp_action::ResultCode::ABORTED:
            RCLCPP_WARN(
                this->get_logger(),
                "Goal was ABORTED (preempted by another goal)! Partial distance: %.2fm",
                result.result->total_distance);
            break;
        default:
            RCLCPP_ERROR(this->get_logger(), "Goal terminated with unknown status code.");
            break;
    }
}

}  // namespace action_cpp_pkg

RCLCPP_COMPONENTS_REGISTER_NODE(action_cpp_pkg::TurtleActionClientNode)
